Storage Refactor: Replication + Custom Storage Support

ikreymer opened the pull request:
Refactors storage to support replicas + custom storages on the Org.
There is a default primary + replica storage, while an Org can also have primary and replica storages.
StorageRef object is used to store references to default and custom storage.

CrawlFile has been updated to contain a StorageRef instead of a def_storage_name, which references
either a default storage (in StorageOps) or custom storage (in Organization)
There is also a 'replicas' Optional[List[StorageRef]] which contains replicas, if any.

CrawlFileOut will contain a numReplicas for how many replicas exist for a given file.

Still need to add migration for def_storage_name -> StorageRef


Part of #1262

ikreymer committed:
work

ikreymer committed:
new org storage model for replicas and custom storages:
- org 'storage' is now a storageref and 'storageReplicas' is a list of storage refs
- a 'StorageRef' can point to either default storage or custom storage
- remove DefaultStorage path, always oid
- org also has a dict of customStorages, name->S3Storage for custom user-specified storages
- storage / storageReplicas can point to either a default or custom storage
- compute custom secret name for default and custom storages in crawlmanager
- api: POST /customStorages - add new customStorage
- api: DELETE /customStorages/{id} - remove existing customStorage
- api: POST /storage - update storage/storageReplicas pointing to existing storage options
- api: GET /storages - list of default and custom storage names

ikreymer committed:
add migration

ikreymer committed:
switch StorageRef { name: str, custom: bool } setup
add migration to migrate to new data layout
add verification of org storages on load

ikreymer committed:
lint fix, add api to get current storage config

ikreymer committed:
check invalid orgs via db queries

ikreymer committed:
store 'storage_secret' seperately from storage_name in configmap,
pass to crawljobs and profilejobs

ikreymer committed:
additional refactor:
- only keep storageName in configmap, compute storagePath + storage secret in operator
- compute prefix paths for both default and custom storages
- fix upload paths
- begin marking default primary and default replicas in helm chart values

ikreymer committed:
reformat

ikreymer committed:
more refactoring:
- move storage prefix checks to StorageRef
- keep track of default primary storage in StorageOps
- use default primary storage for creating new orgs

ikreymer committed:
cleanup: don't set storage on new org creation, will change later

ikreymer committed:
update values, set 'is_default_primary' on tests only

ikreymer committed:
add utility function to get replicas for org

ikreymer committed:
more cleanup / consistent templates

ikreymer committed:
models: CrawlFile & ProfileFile derive from BaseFile

ikreymer committed:
test: change order of default storage

ikreymer committed:
tests: add tests for changing storage to different predefined storage

ikreymer committed:
lint and typo fixes

ikreymer committed:
typo fix

ikreymer committed:
cleanup

ikreymer committed:
cleanup: move signing data out of storage secrets, into seperate signing-secret,
enabled if signing is enabled

ikreymer committed:
Merge branch 'main' into replica-and-custom-storage

ikreymer committed:
- add 'STORE_ENDPOINT_NO_BUCKET_URL' to secret / 'store_endpoint_no_bucket_url' to Storage model to get the endpoint-only without bucket
- add 'replicas' to StorageRef for referencing replicas for gievn file

ikreymer committed:
ci: disable metrics-server for now

ikreymer committed:
Merge branch 'main' into replica-and-custom-storage

ikreymer committed:
remove unused param from create_org_for_user
add 'numReplicas' to CrawlFileOut to report number of replicas on the file

tw4l commented on .github/workflows/k3d-ci.yaml:
```diff
@@ -26,7 +26,7 @@ jobs:
             --agents 1
             --no-lb
-            --k3s-arg "--disable=traefik,servicelb@server:*"
+            --k3s-arg "--disable=traefik,servicelb,metrics-server@server:*"
```
We disabled this for tests, yeah? Maybe a commit from before we did so?

ikreymer replied:
yes, can probably remove, as it'll be merged in before this PR likely.

tw4l commented on backend/btrixcloud/k8sapi.py:
```diff
@@ -148,6 +151,21 @@ async def create_custom_from_dict(self, custom, doc, namespace):
 
+    async def has_storage_secret(self, storage_secret) -> bool:
+        """Check if storage is valid by trying to get the storage secret
+        Will throw if not valid, otherwise return True"""
```
Do we want this to throw an exception? We're using it a few places in crawlmanager without a try/except block so might result in a 500. I guess it is a pretty serious issue if a secret for storage is missing, just wondering if there's a more graceful way to handle it.

ikreymer replied:
Yeah, since this isn't being checked on startup, should just raise an HTTPException

tw4l commented on backend/btrixcloud/migrations/migration_0020_org_storage_refs.py:
```diff
@@ -0,0 +1,44 @@
+"""
+Migration 0020 - Organization Slug
```
```suggestion
Migration 0020 - New Storage Ref System
```

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -370,13 +370,55 @@ class UpdateCrawlConfig(BaseModel):
+
+    def __init__(self, *args, **kwargs):
+        if args:
+            if args[0].startswith("cs-"):
```
This convention seems fine but we'll need to update the docs accordingly.

ikreymer replied:
FWIW, this is for internal use with secret names. The StorageRef stores the name/custom fields so more semantically meaningful, however, need to get a unique name for secrets, so can also be serialized to/from a single string in this way..

tw4l replied:
Ah, right! Makes sense. Thanks for the clarification :)

tw4l commented on backend/btrixcloud/operator.py:
```diff
@@ -143,9 +142,6 @@ class PodInfo(BaseModel):
     newCpu: Optional[int] = None
     newMemory: Optional[int] = None
 
-    # newAllocated: PodResources = PodResources()
-    # force_restart: Optional[bool] = Field(default=False, exclude=True)
```
Does this need to be kept?

ikreymer replied:
No, haven't been using these, so was just a cleanup attempt

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -172,22 +177,48 @@ async def create_default_org(self, storage_name="default"):
         )
         await self.add_org(org)
 
+    async def check_all_storages(self, storage_ops):
```
This seems to be checking only the default storages - should we be checking custom ones too? Or are we not there yet?

ikreymer replied:
Updated to check_all_org_default_storages and updated comment for clarity.

tw4l commented on backend/btrixcloud/profiles.py:
```diff
@@ -62,33 +63,26 @@ async def create_new_browser(
+        # if self.shared_profile_storage:
+        #    storage_name = self.shared_profile_storage
+        #    storage_path = ""
+        # else:
```
Seems like maybe we should either get this working or deprecate it

ikreymer replied:
yeah, i think can remove this, was an early attempt at putting all profiles from across different orgs in same place, but no good reason to keep it now.

ikreymer replied:
Removed, as this was not well designed so far. We may want to think about if want to have different storage for profiles vs crawl data, and if so, maybe allow configuring them separately. But, I think that can be something to revisit in a future PR.

tw4l commented on backend/btrixcloud/storages.py:
```diff
@@ -85,40 +117,126 @@ def _create_s3_storage(self, storage: dict[str, str]) -> S3Storage:
+                raise HTTPException(status_code=400, detail="storage_in_use")
+
+        # pylint: disable=fixme
+        # TODO: also check if any active crawls running with this storage?
```
Bit of an edge case but not a bad idea. Might be better to address this now.

ikreymer replied:
Added! Now checking if crawljobs or profilejobs with label `btrix.storage={storage_name}` exist, before attempting to delete the custom storage.

tw4l commented on backend/btrixcloud/storages.py:
```diff
@@ -184,41 +302,65 @@ async def verify_storage_upload(self, storage: S3Storage, filename: str) -> None
+            s3storage = self.default_storages.get(ref.name)
+        elif not org.storage:
+            raise KeyError(
+                "Referencing custom org storage: {ref.name}, but no custom storage found!"
```
```suggestion
                f"Referencing custom org storage: {ref.name}, but no custom storage found!"
```

tw4l commented on backend/btrixcloud/storages.py:
```diff
@@ -556,13 +695,36 @@ def init_storages_api(org_ops, crawl_manager, user_dep):
+    ):
+        return await storage_ops.remove_custom_storage(name, org)
+
+    # pylint: disable=bare-except, raise-missing-from
```
```suggestion
```

tw4l commented on backend/btrixcloud/storages.py:
```diff
@@ -556,13 +695,36 @@ def init_storages_api(org_ops, crawl_manager, user_dep):
+    ):
+        return await storage_ops.update_storage_refs(storage, org)
+
     # pylint: disable=bare-except, raise-missing-from
```
```suggestion
```

tw4l commented on chart/test/test.yaml:
```diff
@@ -22,7 +22,7 @@ crawler_extra_memory_per_browser: 256Mi
   # specify either username + password (for local mongo)
   username: root
-  password: PASSWORD@
+  #password: PASSWORD@
```
Curious about this, was this from you testing against a live mongo instance?

ikreymer replied:
Yes, this was unintentional!

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -370,13 +370,55 @@ class UpdateCrawlConfig(BaseModel):
     hash: str
     size: int
-    def_storage_name: Optional[str]
+    storage: StorageRef
```
This needs to be added to the migration - without it, I'm seeing this after migrating when attempting to view/download crawl files in the UI:

```
pydantic.error_wrappers.ValidationError: 1 validation error for Crawl
files -> 0 -> storage
  field required (type=value_error.missing)
```

ikreymer replied:
Migration added!

ikreymer committed:
Update backend/btrixcloud/migrations/migration_0020_org_storage_refs.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
migration: add migration from def_storage_name -> StorageRef

ikreymer committed:
add migration for profilefiles as well

ikreymer commented on .github/workflows/k3d-ci.yaml:
```diff
@@ -26,7 +26,7 @@ jobs:
             --agents 1
             --no-lb
-            --k3s-arg "--disable=traefik,servicelb@server:*"
+            --k3s-arg "--disable=traefik,servicelb,metrics-server@server:*"
```
```suggestion
            --k3s-arg "--disable=traefik,servicelb@server:*"
```

ikreymer committed:
undo k3d change here

ikreymer committed:
Update backend/btrixcloud/storages.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
crawlfiles: add optional crc32
- use crc32 when streaming file as part of nested wacz
- use different version stream-zip which allows empty crc32 (for now)

ikreymer committed:
storage secret check: raise HTTPException

ikreymer committed:
rename check_all_org_default_storages and update comment

ikreymer committed:
undo test change

ikreymer committed:
move raise-missing-from to top of file

ikreymer committed:
more remove unused pylint disables

ikreymer committed:
remove 'shared_profile_storage' option, no longer supported in new setup
was only used for testing, and not well designed

ikreymer committed:
add support for checking if storage secret is in use before attempting to remove:
- add 'btrix.storage' labels to crawljobs and profilejobs
- add has_custom_jobs_with_label() to check custom jobs
- check if crawljobs or profilejobs with matching storage label exist before attempting to remove the storage
- use more consistent labeling for crawljobs / profilejobs

ikreymer commented:
@tw4l I think this should be ready for re-review. In particular, the migration should now be working. There's definitely more tests that could be added around custom storages as well.
We could potentially disable those endpoints for now to reduce surface area, and come back to them later, with the architecture in place.

tw4l commented:
> We could potentially disable those endpoints for now to reduce surface area, and come back to them later, with the architecture in place.

Yes I think that'd be a good idea until we're ready to support them! Could just comment them out for now.

tw4l committed:
Merge branch 'main' into replica-and-custom-storage

tw4l commented on backend/btrixcloud/migrations/migration_0020_org_storage_refs.py:
```diff
@@ -0,0 +1,79 @@
+
+        # CrawlFile Migrations
+        mdb_crawls = self.mdb["crawls"]
+        async for crawl in mdb_crawls.find({"files.def_storage_name": {"$ne": None}}):
```
I think we need to run this on all files (crawl and profile), even if `def_storage_name` is `None`. Otherwise, we get the following error when trying to list profiles as an example:

```
  File "/app/btrixcloud/profiles.py", line 229, in list_profiles
    profiles = [Profile.from_dict(res) for res in results]
  File "/app/btrixcloud/profiles.py", line 229, in <listcomp>
    profiles = [Profile.from_dict(res) for res in results]
  File "/app/btrixcloud/db.py", line 214, in from_dict
    return cls(**data)
  File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for Profile
resource -> storage
  field required (type=value_error.missing)
```

tw4l replied:
Profile files at least made in current main have `def_storage_name` set to `None` but still need `storage` added in migration.

tw4l committed:
Migrate all crawl and profile files

Also fixes as few bugs, such as setting profile.resources instead
of profile.resource

tw4l committed:
Add pylint disable command

ikreymer committed:
disable setting custom storage for now, until have chance to test better / fully support

ikreymer committed:
disable custom storage test

tw4l committed:
Fix files migration
